### PR TITLE
[FEATURE] Add performance test

### PR DESCRIPTION
**Pull request by jdimatteo**
Add performance benchmarks to serve two purposes:

1. Quantify the impact of PRs that are intended to improve performance, and 
2. Measure trends over time to identify/prevent performance regressions.

Changes proposed in this pull request:
- Add performance benchmark using BigQuery to test between 1 and 100 tables (currently taking between ~7 seconds and ~12 minutes).

Please note that there are 2 (currently incomplete) follow on PRs: https://github.com/great-expectations/great_expectations/pull/3285 is intended to integrate these tests into CI, and https://github.com/great-expectations/great_expectations/pull/3287 leverages these tests to show performance improvements (e.g. 9.9x faster) with concurrency.

### Configuring Data Before Running Performance Tests

These performance tests use BigQuery.

Before running a performance test, setup data with `setup_bigquery_tables_for_performance_test.sh`.

For example:

```bash
GE_TEST_BIGQUERY_PEFORMANCE_DATASET=<YOUR_GCP_PROJECT> setup_bigquery_tables_for_performance_test.sh
```

Please see also the [contributing BigQuery tests documentation](https://docs.greatexpectations.io/docs/contributing/contributing_test#bigquery-tests) for more info on how to configure BigQuery to run these tests.

### Running the Performance Tests

Run the performance tests with pytest, e.g.

```
pytest tests/performance/test_bigquery_benchmarks.py \
  --bigquery --performance-tests \
  -k 'test_taxi_trips_benchmark[1-True]'  \
  --benchmark-json=tests/performance/results/`date "+%H%M"`_${USER}.json \
  --no-spark --no-postgresql -rP -vv
```

Some benchmarks take a long time to complete. In this example, only the relatively fast `test_taxi_trips_benchmark[1-True]` benchmark is run and the output should include runtime like the following:

```
--------------------------------------------------- benchmark: 1 tests ---------------------------------------------------
Name (time in s)                         Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------
test_taxi_trips_benchmark[1-True]     5.0488  5.0488  5.0488  0.0000  5.0488  0.0000       0;0  0.1981       1           1
--------------------------------------------------------------------------------------------------------------------------
```

The result is saved for comparisons as described below.

### Comparing Performance Results

Compare test results in this directory with `py.test-benchmark compare`, e.g.

```
$  py.test-benchmark compare --group-by name tests/performance/results/initial_baseline.json tests/performance/results/*${USER}.json                                                                   

---------------------------------------------------------------------------- benchmark 'test_taxi_trips_benchmark[1-True]': 2 tests ---------------------------------------------------------------------------
Name (time in s)                                        Min               Max              Mean            StdDev            Median               IQR            Outliers     OPS            Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_taxi_trips_benchmark[1-True] (initial_base)     5.0488 (1.0)      5.0488 (1.0)      5.0488 (1.0)      0.0000 (1.0)      5.0488 (1.0)      0.0000 (1.0)           0;0  0.1981 (1.0)           1           1
test_taxi_trips_benchmark[1-True] (2114_work)        6.4675 (1.28)     6.4675 (1.28)     6.4675 (1.28)     0.0000 (1.0)      6.4675 (1.28)     0.0000 (1.0)           0;0  0.1546 (0.78)          1           1
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```

Please refer to [pytest-benchmark documentation](https://pytest-benchmark.readthedocs.io/en/latest/comparing.html) for more info.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Comment by jdimatteo** [comment]
The failing test doesn't appear related to this PR: https://dev.azure.com/great-expectations/great_expectations/_build/results?buildId=14905&view=logs&j=459c5136-8c6f-5453-4213-f6ede6e2bb47&t=d6598094-8f5d-515a-f045-7e469107b3ef&l=39

**Comment by talagluck** [comment]
Thanks so much for this contribution, @jdimatteo ! We will review and discuss internally, and come back with some feedback later this week.

File: docs/contributing/contributing_test.md [hunk 1]
```diff
@@ -149,6 +149,63 @@ Note: If you are implementing a new Expectation, but don’t plan to immediately
...
+For more information on getting started with BigQuery, please refer to the [above section on BigQuery tests](#bigquery-tests).
+
+##### Running the Performance Tests
+
+Run the performance tests with pytest, e.g.
+
+```
+pytest tests/performance/test_bigquery_benchmarks.py \
+  --bigquery --performance-tests \
+  -k test_bikeshare_trips_benchmark[1] \
```
Comment (Shinnnyshinshin) [inline docs/contributing/contributing_test.md]: for this flag, pytest actually required this for me, with the single quotes and `True`

```bash
  -k 'test_bikeshare_trips_benchmark[1-True]' 
```

**Review by jdimatteo — COMMENTED** [review]
Beautiful thorough improvements @Shinnnyshinshin  -- thanks so much for your commits cleaning up this code and updating to use the standard NYC taxi data!

**Review by Shinnnyshinshin — APPROVED** [review]
Approving. Thank you very much for this @jdimatteo

**Pull request merged**